File: 007 有序表的合并/使用之前的链表实现.swift
import Foundation

func merge(_ listA: LinkList<Int>, _ listB: LinkList<Int>) -> LinkList<Int> {
    let listC = LinkList<Int>()
    
    let lenA = listA.length
    let lenB = listB.length
    
    var idxA: UInt = 0
    var idxB: UInt = 0
    
    while idxA < lenA && idxB < lenB {
        let valA = listA.getValue(of: idxA)!
        let valB = listB.getValue(of: idxB)!
        
        if valA < valB {
            _ = listC.append(valA)
            idxA += 1
        } else {
            _ = listC.append(valB)
            idxB += 1
        }
    }
    
    while idxA < lenA {
        listC.append(listA.getValue(of: idxA)!)
        idxA += 1
    }
    
    while idxB < lenB {
        listC.append(listB.getValue(of: idxB)!)
        idxB += 1
    }
    
    return listC
}

// 测试数据
var listA = LinkList<Int>()
var listB = LinkList<Int>()

var arrA = [Int]()
var arrB = [Int]()

for _ in 0..<6 {
    arrA.append(Int.random(in: 0..<30))
}

for _ in 0..<9 {
    arrB.append(Int.random(in: 0..<90))
}

for v in arrA.sorted() {
    _ = listA.append(v)
}

for v in arrB.sorted() {
    _ = listB.append(v)
}

print("数组A：")
listA.traverse()
print("数组B：")
listB.traverse()

print("合并后数组：")
let listC = merge(listA, listB)
listC.traverse()

File: 007 有序表的合并/使用之前的链表实现.c
#include <stdio.h>
#include <stdlib.h>

void * merge(LinkList *listA, LinkList *listB) {
    int lenA = length(listA);
    int lenB = length(listB);
    
    LinkList *listC = createList();
    
    int idxA = 0;
    int idxB = 0;
    
    while (idxA < lenA && idxB < lenB) {
        int valA, valB;
        getValue(listA, idxA, &valA);
        getValue(listB, idxB, &valB);
        if (valA <= valB) {
            append(listC, valA);
            idxA++;
        } else {
            append(listC, valB);
            idxB++;
        }
    }
    
    while (idxA < lenA) {
        int val;
        getValue(listA, idxA, &val);
        append(listC, val);
        idxA++;
    }
    
    while (idxB < lenB) {
        int val;
        getValue(listB, idxB, &val);
        append(listC, val);
        idxB++;
    }
    return listC;
}

int main() {
    LinkList *listA = createList();
    LinkList *listB = createList();
    
    int arrA[4] = {21, 32, 36, 88};
    int arrB[7] = {2, 15, 37, 99, 123, 345, 432};
    
    for (int i = 0; i < 4; i++) {
        append(listA, arrA[i]);
    }
    
    for (int i = 0; i < 7; i++) {
        append(listB, arrB[i]);
    }
    
    printf("数组A：");
    traverse(listA);
    printf("数组B：");
    traverse(listB);
    
    LinkList *listC = merge(listA, listB);
    printf("合并后数组：");
    traverse(listC);
    
    return 0;
}
